django/contrib/admin/templatetags/admin_modify.py

```diff
@@ -44,25 +44,38 @@ def submit_row(context):
     """
     Displays the row of buttons for delete and save.
     """
+
+    opts = context['opts']
+    add = context['add']
     change = context['change']
     is_popup = context['is_popup']
     save_as = context['save_as']
     show_save = context.get('show_save', True)
     show_save_and_continue = context.get('show_save_and_continue', True)
+    has_change_permission = context['has_change_permission']
+    has_add_permission = context['has_add_permission']
+    has_editable_inline_admin_formsets = context['has_editable_inline_admin_formsets']
+
+    can_save = (has_change_permission and change) or (has_add_permission and add) or has_editable_inline_admin_formsets
     ctx = Context(context)
     ctx.update({
+        'opts': opts,
         'show_delete_link': (
             not is_popup and context['has_delete_permission'] and
             change and context.get('show_delete', True)
         ),
-        'show_save_as_new': not is_popup and change and save_as,
+        'show_save_as_new': not is_popup and has_change_permission and change and save_as,
         'show_save_and_add_another': (
-            context['has_add_permission'] and not is_popup and
-            (not save_as or context['add'])
+            has_add_permission and not is_popup and
+            (not save_as or add) and can_save
         ),
-        'show_save_and_continue': not is_popup and context['has_change_permission'] and show_save_and_continue,
-        'show_save': show_save,
+        'show_save_and_continue': not is_popup and can_save and show_save_and_continue,
```

Comment:
This looks OK on manual testing. 

* I have a user with `add` and `view` permissions on X
* I add an X, using the "Save and continue" button. 
* I'm redirected to the same page to view the object. Only option is "Close without saving".

Reply:
Except the message displayed isn't quite right in this case: 

```
The X "Great new X" was added successfully. You may edit it again below.
```

It's not true that it may be edited.

Reply:
Thinking about this again, my point was that "Save and continue" doesn't make sense if you **only** have the *add* permission without *view* or *change* also.

Before you had to have the *change* permission. Now it ought to be *change* or *view*, but has been changed to use `can_save` which allows just having the *add* permission.

This makes no sense. Either we then have to redirect somewhere else other than the change/view view making the behaviour different, i.e. the same as the "Save" button, or we redirect and get a permission denied error.

I haven't tested these scenarios, but just by inspection of the code, it is clear we are changing the expected behaviour here.